### ISSUE TITLE
Top "Scores" [$125]

### ISSUE THREAD

**Issue by kevingranade** [CleverRaven/Cataclysm-DDA#4173]
Suggested on whalesdev, a per-world or per game instance list of performance of various characters, in various categories like furthest traveled, most explored, most monsters killed, longest life, (shortest life, heh).
An interesting complement to the memorial file.
##

**Comment by Chezzo** [comment]
I would really like to see this issue done, so I put $35 more on it. (And wubblemeister put $50! You rule @wubblemeister!)

I think this would be so awesome for @r7st's SSH server at watchcdda.com, both for regular play and multiplayer. And for comparing to everyone else on the internet. "I travelled 5000 miles!" "Whoa, rad, I explored 8000 tiles!"

I think could use some sort of multiplier for spawn rate, multipool, and wander spawn. And a general score that combines the things in the op.

Some other things that would be fun to add would be:

Bionics Added
Mutations Gained
Drugs Taken
NPCs/pets/zlaves
Mended limbs
Books Read

**Comment by Wuzzy2** [comment]
Quoting from #23436:

> It doesn't need to be fancy, a simple ranking based on the time you survived would be a good start.
Just show rank, player name and survival time in the list.
> 
> But to be fair, there need to be some rules:
> 
>  * There is no highscore for the freeform pool as this is just a sandbox
>  * As soon you use ANYTHING in the debug menu (e.g. spawn item), you're disqualified from a highscore (well, for the current character, at least). This has to persist in the savegame, obviously
>  * If any gameplay-related setting in the debug settings has a non-default value (e.g. maximum character points = 1000), you're disqualified, too
> 
> To avoid main menu clutter, the highscore list could go into the “Specials” menu.
> 
> There's just a “little” problem. The highscores are not really comparable because there is a ton of other things you can change. Mods may change gameplay considerably and also the world settings may make a game much easier.
> 
> One partial solution to this would be only worlds with default mods and world settings are eligible. I don't have a better idea right now.

**Comment by Shoes01** [comment]
Regarding the little problem: save all the settings too. That way when people compare high scores, then can also confess the settings they used to get it.

**Comment by Wuzzy2** [comment]
Saving *all* settings is clearly overkill. Only gameplay-related stuff should ever be saved.

But even then there's still a “tiny” problem: Players can change gameplay settings at will in-game.

**Comment by Shoes01** [comment]
Have a settings change, like accessing the debug menu, invalidate the high score. Or not ;p the scores will be a plain text anyway!

**Comment by Wuzzy2** [comment]
Debug menu is not the settings menu. Of course any usage of the debug menu's features should instantly disqualify you.

Many settings are irrelevant for scoring, like screen size, sidebar style, tiles mode and other graphics-related things.

**Comment by Epictyphlosion** [comment]
I'd love to see something like this. I have no idea how it would be coded in, but if someone does, it's worth almost $100 if you do!

**Comment by Asmageddon** [comment]
Tracking the actual stats sounds straightforward-ish, though it's gonna touch a ***lot*** of code meaning a lot of effort, plus if anyone ever breaks it they're unlikely to notice... 

Bigger issues are... how exactly do you compile those into a score? That sounds almost impossible to do in a way that isn't abusable. Even "difficulty multiplier" is a bit goofy because once you're out of the earlygame, most of the settings don't affect you as much anymore.

**Comment by jbytheway** [comment]
Since @Leland pointed it out in #33369, I will mention that I am trying to work towards this.  My overall plan is roughly as follows:
* Add event bus so that the events relevant for scoring can be tracked without lots of code caring specifically about scoring (#32891).
* Convert some existing game features to use the event bus where that makes sense (kill tracker already implemented (#33369), next is the memorial file).
* Add a new statistics tracking component that takes all the events and turns them into suitable summary statistics from which scores might be computed.
* Allow scores and achievements to be defined in json in terms of these statistics.
* Provide a way to explore the scores / achievements from your past games.  As far as I can tell watchcdda no longer exists, but of course it would be easy to provide machine-readable access to scores for such a service to compile high scores tables or whatever else they might wish.

Scores are cool, but I think achievements are actually much more valuable for enhancing gameplay, and can be derived from the same infrastructure.  I'm sure the community can come up with plenty of fun achievements to help people discover new aspects of the game that they might not otherwise explore, and give direction and purpose to a game that can, to some people, seem open-ended to an intimidating degree.  And if all goes to plan they'll all be defined in json, so people can easily experiment with new score / achievement definitions in mods to figure out what makes sense.

**Comment by Maleclypse** [comment]
@jbytheway Is there anything remaining to do in this issue to warrant leaving it open?

**Comment by jbytheway** [comment]
Yes. We still don't have an actual top scores window. I'm planning to work on that once the release is done.

**Comment by Leland** [comment]
Threw another $15 on it for ya! Excited to see this land  👏

**Comment by kevingranade** [comment]
Well we got achievements out of this but that top scores window doesn't seem like it's going to happen.

**Comment by Asmageddon** [comment]
Probably for the better if you ask me. Games like CDDA should stand on the intrinsic fun of the process of playing, not on chasing high scores.
